rfcs/constructor-unboxing.md

```diff
@@ -0,0 +1,169 @@
+# Constructor unboxing
+
+### Motivating example: compact rope representation
+
+Data structures defined in OCaml are often less compact than they might be, because of boxing.
+
+For example, here is a type for representing [ropes](https://onlinelibrary.wiley.com/doi/abs/10.1002/spe.4380251203):
+
+```ocaml
+type rope = Leaf of string
+          | Branch of { llen: int; l:rope; r:rope }
+```
+
+With this definition the value `Branch {llen=3; l=Leaf "abc"; r=Leaf "def"}` has the following representation:
+
+
+```
+[--B|-3-|-∘-|-∘-]
+         /     \
+        /       \
+       /         \
+      [--L|-∘-]   [--L|-∘-]
+           /             \
+          /               \
+         "abc"             "def"
+```
+
+In the general case each part of this representation serves a purpose.  For example, in order to distinguish `Leaf` nodes from `Branch` nodes at run-time, each constructor is represented by a tagged block.  However, for this particular data type the block representing `Leaf` nodes (`[--L|-∘-]`) is unnecessary; since strings are already distinguishable from other blocks, the value could in principle be represented more compactly:
+
+
+```
+[--1|--3|-∘-|-∘-]
+         /     \
+        /       \
+       /         \
+       "abc"      "def"
+```
+
+
+### The basic idea 
+
+Adding an `[@@unboxed]` annotation to a variant definition indicates that the representation of unary constructors should not involve an additional block:
+
+```ocaml
+type rope = Leaf of string
+          | Branch of { llen: int; l:rope; r:rope } [@@unboxed]
+```
+
+Only a subset of variant definitions support `[@@unboxed]`.  In particular, it must be possible to distinguish the arguments of unary constructors from each other (and from constant constructors in the same definition) at run-time.  For example, the following definition is not allowed, since the arguments of `X` and `Y` have the same representation:
+
+```ocaml
+type strings = X of string | Y of string [@@unboxed] (* Invalid! *)
```

Comment:
I'm not sure about the original specification "should not involve an additional block", and what it is supposed to reject.
For instance, for the following
```ocaml
type 'a or_int = Ret of 'a | Err of int [@@unboxed]
```
it is possible to unbox `Err` but clearly not `Ret`, so will it be accepted?

In general, should `[@@unboxed]` be seen as prescriptive or suggestive.
If it is prescriptive, wouldn't it be better to annotate each constructor individually, meaning that both
```ocaml
type strings = X of string [@@unboxed] | Y of string
```
and
```ocaml
type strings = X of string | Y of string [@@unboxed]
```
would be valid, but not
```ocaml
type strings = X of string [@@unboxed] | Y of string [@@unboxed]
```

Otherwise, it could be just suggestive, meaning "attempt to flatten the type as much as possible".
We might still want a warning when nothing is done.